Move middlewares to sagas

superhawk610 opened the pull request:
**Related PR:** #97, #111, #117, #122

**Summary:**
In the words of @mik639,
> redux-saga helps to write flat sync code, and it also has one more advantage it's pretty testable :)

Instead of watching for actions in the middle of the redux chain, redux-saga allows all side-effects to be managed independently at the end of the chain. Sagas leverage ES6 generators to allow for writing code that's both easy to read and easy to test.

This PR adds redux-saga, complete with flow typing, and replaces the 3 existing middlewares with 3 fully-tested sagas. It also includes a few minor tweaks to existing selectors and tests to account for the new syntax. Additional minor changes include moving `projects.reducer.test.js` up a level from the `tests` directory to live alongside its reducer, and adding a couple missing assertions in `create-project.service.test.js` (as well as renaming it from `create-project-service.test.js` to conform to existing naming conventions).

superhawk610 commented:
I didn't realize we already had `chalk@2.3.0` as a dependency. I can revert the version if needed.

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+
+        const isError = text.includes('Failed to compile');
+
+        emitter({ channel: 'stdout', text, isError });
```
What do you think about getting rid of the `isError` property, but instead just emitting on `stderr` if the `stdout` _should_ be an error.

We can use this little abstraction to standardize the usage of out/err, by our own measures of what should/shouldn't be an error.

On the other hand, it's a bit confusing that depending on where you put your `console.log`, the same caught output could be on a different channel...

I think it's worth that small bit of added complexity, because right now it's a bit weird that there's an `isError` prop.

What do you think? Am I missing something?

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+
+    yield call([ipcRenderer, ipcRenderer.send], 'addProcessId', child.pid);
+
+    const stdioChannel = createStdioChannel(child, {
```
Interesting! Never worked with channels in Redux Saga, seems useful though!

joshwcomeau replied:
Hm, looking at [the docs](https://github.com/redux-saga/redux-saga/blob/master/docs/advanced/Channels.md#using-the-eventchannel-factory-to-connect-to-external-events), it seems like we're supposed to `yield call` the channel-creator:

```js
const stdioChannel = yield call(createStdioChannel, child, { ... });
```

Because `call` will work synchronously when it isn't calling a generator or a function that returns a promise, it seems like it works the same way, except I think it returns an effect description instead of the actual channel.

I tried making this change locally, and it still works the same, although it causes 3 tests to fail 😬

Asking for my own education (I'm still reacquainting myself with redux saga), not because I'm requesting any changes.

joshwcomeau replied:
Yeah, makes sense. I've written some saga tests since writing that comment and it's clearer to me now. I think this is good :+1:

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+
+    return () => {
+      /* unsubscribe any listeners */
+    };
```
Do we need to do this? Since we don't have any subscribed listeners, is it valid to just `return null` here?

superhawk610 replied:
> Error: in eventChannel: subscribe should return a function to unsubscribe

redux-saga throws if a channel creator doesn't return a function ¯\\__(ツ)_/¯

superhawk610 replied:
I'll add a comment to this effect.

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+    // at bare minimum we expect to have 2 messages queued at some point (as
+    // the exit channel completes, it should emit the return code of the process
+    // and then immediately END
+  }, buffers.expanding(2));
```
Curious why we need to specify a buffer at all? If I'm reading the docs correctly, the default is to have an unlimited buffer size. It seems like `buffers.expanding` can also grow to infinite size, we're just setting an initial size?

It's very possible (likely, even) that I'm misunderstanding something though, so please let me know :)

superhawk610 replied:
> Like actionChannel (Effect), eventChannel (a factory function, not an Effect) creates a Channel for events but from event sources other than the Redux Store.

> Note: messages on an eventChannel are not buffered by default. You have to provide a buffer to the eventChannel factory in order to specify buffering strategy for the channel (e.g. eventChannel(subscriber, buffer)). [See the API docs](https://github.com/redux-saga/redux-saga/tree/master/docs/api#buffers) for more info.

Since this isn't processing redux actions, I went with an `eventChannel`, which doesn't buffer by default. In this particular case we probably don't need a buffer (?) since the console output is handled immediately without any blocking asynchronous calls, but I wasn't confident of that, and should we ever implement some sort of async handlers for console output down the road then this pattern would be required.

I'll remove it since I don't believe it's necessary now and just leave in a comment about how it should be used in the future.

joshwcomeau replied:
ahh I see.

Yeah, I could go either way, but I'm tempted to ditch it for now since it isn't needed, and "buffers" is scary-sounding haha.

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+      }
+    } finally {
+      /* child process has completed */
+    }
```
I think we can omit the `finally` block, unless there are short-term plans to use it?

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+        .includes('Are you sure you want to eject? This action is permanent');
+
+      if (isEjectPrompt) {
+        sendCommandToProcess(child, 'y');
```
so glad we can do this now :D I know it's unrelated to this change, it just makes me really happy, for the future test possibilities

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+    ? 'Server stopped'
+    : 'Task aborted';
+
+  yield put(receiveDataFromTaskExecution(task, chalk.bold.red(abortMessage)));
```
Thanks for using Chalk here!

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+  yield put(receiveDataFromTaskExecution(task, chalk.bold.red(abortMessage)));
+}
+
+export function* displayTaskComplete(task: Task): Saga<void> {
```
Looks like this method should also receive a `wasSuccessful` boolean, and then use that boolean to determine what the message should be ("Task completed" or "Task failed").

Tried this locally and indeed, I'm getting "Task complete" when the task fails:

<img width="670" alt="screen shot 2018-08-24 at 8 34 24 am" src="https://user-images.githubusercontent.com/6692932/44584918-87bc9f80-a778-11e8-9cec-d983f63ca9ce.png">

joshwcomeau commented on src/sagas/task.saga.js:
```diff
@@ -0,0 +1,339 @@
+    window.process.env.PATH +
+    path.delimiter +
+    path.join(projectPath, 'node_modules', '.bin'),
+});
```
Feels like this method should move somewhere... maybe `platform.service.js`?

superhawk610 replied:
Yeah that definitely makes more sense.

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,382 @@
+
+// we want to watch and ensure that caught errors
+// are logged to the error log
+global.console = { error: jest.fn() };
```
I noticed when running the tests that `console.log` fails ("console.log is not a function").

It also kinda worries me to do this, since it's never getting unset; I can imagine it causing problems for other tests run in the same session.

Maybe we can do this in a `beforeAll` block, and then unset it in an `afterAll`?

```js
// At the top, here:
const consoleOriginal = global.console;

// Within a test block:
beforeAll(() => {
  global.console = {
    ...consoleOriginal, // I wonder if this would work, to preserve `console.log`?
    error: jest.fn()
  };
});

afterAll(() => {
  global.console = consoleOriginal;
})
```

superhawk610 replied:
It's only used in this single test, so I just moved it all to that block. There's also no need to replace `global.console` entirely when we can just replace `global.console.error`. Good catch.

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,382 @@
+
+describe('task saga', () => {
+  beforeEach(() => {
+    window.process = { env: {} };
```
Right, the idea is that the saga uses `window.process.env`, and that's technically undefined in the tests?

This might be a bit paranoid, but I wonder if we should do this conditionally:

```
if (!window.process) {
  window.process = { env: {} };
}
```

This way, if we ever move test runners, and the runner provides a window.process, we aren't clobbering it unconditionally?

Also, I think this makes more sense as a `beforeAll`? Shouldn't need to re-run for every test.

Finally (sorry for all this!), would be great to add a comment pointing out why this is needed (looks like it's for `getBaseProjectEnvironment`?)

superhawk610 replied:
All good points. Changed to `beforeAll` and made conditional.

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,382 @@
+      // `loadDependencyInfoFromDisk` is a thunk, thus two copies
+      // constructed with identical arguments will return unique
+      // copies of the same anonymous function. As such, their JSON
+      // stringified representations are used for testing equality.
```
I think `JSON.stringify` is stripping out the functions altogether; when I log it, I get `"{\"@@redux-saga/IO\":true,\"PUT\":{\"channel\":null}}"`. No mention of the `action` created.

I think that's OK, though; ideally I want to get rid of redux-thunk altogether, and move it to a saga, so this is a temporary problem :D

j-f1 replied:
Yep, `JSON.stringify` skips keys with function values.

superhawk610 replied:
Yeah, this is one spot where redux-saga and redux-thunk don't play well together. I'll add a TODO to update these once redux-thunk is out of the picture.

joshwcomeau replied:
Happily there's only 1 thunk left, and it should be gone soon!

joshwcomeau commented on package.json:
```diff
@@ -44,6 +44,7 @@
     "fix-path": "2.1.0",
     "gatsby-cli": "1.1.58",
     "ps-tree": "1.1.0",
+    "redux-saga": "0.16.0",
```
Ah I just noticed; this should be a `devDependency`.

While it seems counterintuitive that we're moving all the stuff we actually use to `devDependencies`, the rationale is that we don't want to bundle them in our dist bundles. We only want to ship the `node_modules` that the Node app needs to do its job at runtime, not the ones bundled by Webpack for "browser" usage.

joshwcomeau commented on src/sagas/import-project.saga.js:
```diff
@@ -0,0 +1,173 @@
+        'Unknown error',
+        'An unknown error has occurred. Sorry about that! Details have been printed to the console.'
+      );
+      break;
```
I don't actually think that's true. I've done stuff like this before:

```js
switch (size) {
  case 'small': 
    ...

  default:
  case 'medium':
    ...

  case 'large':
  ...
 
}
```

But yeah, we can kill the `break` here. Just being nitpicky haha.

superhawk610 replied:
Somewhere my 10th grade Java teacher is judging you for that. 😄

I honestly didn't know default could be anything but last, TIL.

superhawk610 commented:
While playing around with the channel structure and `receiveDataFromTaskExecution` action, I realized that Jest outputs some of it's standard logs to `stderr`, which is problematic because it causes the Test UI to flash `Failed` during normal operations:

![jest-stderr](https://user-images.githubusercontent.com/18172185/44760433-bf2aa380-ab0c-11e8-8e48-b811aab60fc0.gif)

joshwcomeau commented:
> While playing around with the channel structure and receiveDataFromTaskExecution action, I realized that Jest outputs some of its standard logs to stderr, which is problematic because it causes the Test UI to flash Failed during normal operations:

Ahh that's not good!

This is hitting up against our flawed way of handling statuses for tasks. I think for short-term tasks like running tests, the logs shouldn't affect the status while the task is running (while it's running, the status should be `idle`, and should only be set based on the end state of the task).

I think your PR is significant enough without trying to tackle this as well. We can merge this with the bug, and you or I can take a stab at fixing this issue afterwards (before releasing)

superhawk610 commented:
Alright, I reverted error handling to its current behavior and added a note.

The test suite passes after rebasing on master, but I'm now running into some issues with `redux-storage` (I'm a bit out of the loop with that so it's possible I'm just misunderstanding how to correctly configure it with an existing install). Even after deleting `guppy-projects-dev` entirely, I can't seem to fully reset the redux cache, so I end up loading in to a blank screen with no projects and have to use the menu item to create a new project. In addition, once I've created a project, no icon appears on the left.

![image](https://user-images.githubusercontent.com/18172185/44829028-9677dc00-abe8-11e8-8379-fd2cb09f4d73.png)

As shown, I'm also seeing some new warnings from Electron about insecure content...?

joshwcomeau commented:
Hi @superhawk610! Sorry for the confusion.

> Even after deleting guppy-projects-dev entirely, I can't seem to fully reset the redux cache, so I end up loading in to a blank screen with no projects and have to use the menu item to create a new project. In addition, once I've created a project, no icon appears on the left.

There appears to be some sort of issue with the onboarding status. Good catch on the `translate`. Investigating now.

This is an issue in master, so it shouldn't delay this review.

> As shown, I'm also seeing some new warnings from Electron about insecure content...?

I've seen these warnings since the very first version @_@ at least, the top 2. Not sure about the last one.

I'm not too worried, since the app doesn't actually make any external requests other than to NPM for dependency search. The insecure content it's talking about is the locally-packaged JS bundle.

We should at some point find a way either to disable the warnings, or to rework our dev flow to satisfy them, since it is rather alarming-looking.

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,393 @@
+    const port = 3000;
+    const { args, env } = getDevServerCommand(task, project.type, port);
+    const processId = 12345;
+    const mockProcess = mockSpawn(processId);
```
Huh, `cloneableGenerator` is neat. I could have used that in my tests, will update at some point.

It's a bit weird to me to see random constants defined in a `describe`, I'm used to seeing this in a `beforeAll` block... but I suppose if it works, there's no problem!

superhawk610 replied:
You'll have to excuse me, I'm still figuring out Jest conventions.

On that note, how does `beforeAll()` work with block-scoped variables? If I do

```js
beforeAll(() => {
  const saga = /* ... */;
});
```

can I then do

```js
it('should do whatever', () => {
  saga.next();
});
```

or do I first need to declare it in the parent scope and set its value in `beforeAll()`?

superhawk610 replied:
Let me know and I'll rewrite these.

joshwcomeau replied:
Yeah, that's right:

```js
let thing;
beforeAll(() => {
  thing = 'yeah';
});
```

That said, I don't think it matters at all. No rewriting necessary :) I've seen it done that way but I don't actually know why (maybe for parallelization somehow?). The way you have it works :+1:

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,393 @@
+      expect(console.error).toBeCalled();
+
+      // restore `console.error` to its original function
+      global.console.error = consoleErrorOriginal;
```
Yay thanks for this.

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,393 @@
+
+      expect(saga.next().value).toEqual(select(getProjectById, task.projectId));
+      // stringify to avoid deep equal inconsistencies from thunk
+      // TODO: remove `JSON.stringify` once `redux-saga` is removed
```
Nit: once `redux-thunk` is removed, not `redux-saga`

joshwcomeau commented on src/sagas/import-project.saga.test.js:
```diff
@@ -0,0 +1,288 @@
+    app: { getAppPath: () => '' },
+    dialog: { showOpenDialog: jest.fn(), showErrorBox: jest.fn() },
+  },
+}));
```
We have a shared electron mock now, in `__mocks__/electron.js`. We should add these new `dialog` methods to it, and remove this mock.

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,393 @@
+  remote: {
+    app: { getAppPath: () => '' },
+  },
+}));
```
Same note here about electron mocking

joshwcomeau commented on src/services/platform.service.js:
```diff
@@ -37,6 +37,17 @@ export const formatCommandForPlatform = (command: string): string =>
+    window.process.env.PATH +
+    path.delimiter +
+    path.join(projectPath, 'node_modules', '.bin'),
+});
```
Good call moving this :+1:

joshwcomeau commented on src/store/index.js:
```diff
@@ -41,14 +38,7 @@ export default function configureStore() {
-        importProjectMiddleware,
-        storageMiddleware,
-        sagaMiddleware
-      ),
```
:D

joshwcomeau commented on src/sagas/task.saga.test.js:
```diff
@@ -0,0 +1,390 @@
+      );
+    });
+
+    describe('running', () => {
```
I had to create this new block because of a `no-shadow` lint error; `saga` was being defined in the parent block, and then shadowed in the earlier tests like `should add --coverage arg to create-react-app tests`.

I think it's better anyway to have the tests that use a "parent" saga be defined in their own "describe". Although "running" is not a great name for that block. Open to suggestions!

superhawk610 replied:
This looks great, and I agree, it does make more sense to group "child" saga tests.
